kore/src/Kore/Builtin/IO.hs

```diff
@@ -74,29 +113,409 @@ symbolVerifiers =
+                    }
+                []
+
 evalLogString :: Builtin.Function
 evalLogString _ _ [] = Builtin.wrongArity logStringKey
 evalLogString _ sort [InternalString_ InternalString{internalStringValue}] = do
-    tools <- askMetadataTools
     infoUserLog internalStringValue
-    return $ Pattern.fromTermLike $ TermLike.mkApplySymbol (dotk tools) []
-  where
-    dotk tools =
-        Symbol
-            { symbolConstructor = ident
-            , symbolParams = []
-            , symbolSorts =
-                ApplicationSorts
-                    { applicationSortsOperands = []
-                    , applicationSortsResult = sort
-                    }
-            , symbolAttributes = Tools.symbolAttributes tools ident
-            }
-    ident = implicitId "dotk"
+    mkDotk sort
 evalLogString _ _ [_] = empty
 evalLogString _ _ (_ : _ : _) = Builtin.wrongArity logStringKey
 
 logStringKey :: IsString s => s
 logStringKey = "IO.logString"
+
+returnErrnoIfMinus1OrRetry :: (Eq a, Num a) => IO a -> IO (Either Errno a)
+returnErrnoIfMinus1OrRetry f =
+    do
+        res <- f
+        if res == -1
+            then do
+                err <- getErrno
+                if err == eINTR
+                    then returnErrnoIfMinus1OrRetry f
+                    else return $ Left err
+            else return $ Right res
+
+parseMode :: Text -> Maybe (OpenMode, OpenFileFlags, Maybe FileMode)
+parseMode = \case
+    "r" -> Just (ReadOnly, defaultFileFlags, Nothing)
+    "r+" -> Just (ReadWrite, defaultFileFlags, Nothing)
+    "w" -> Just (WriteOnly, defaultFileFlags{trunc = True}, Just 0o666)
+    "w+" -> Just (ReadWrite, defaultFileFlags{trunc = True}, Just 0o666)
+    "a" -> Just (WriteOnly, defaultFileFlags{append = True}, Just 0o666)
+    "a+" -> Just (ReadWrite, defaultFileFlags{append = True}, Just 0o666)
+    _ -> Nothing
+
+-- copied from System.Posix.IO.Common as this module is hidden
+open_ ::
+    -- | Pathname to open
+    CString ->
+    -- | Read-only, read-write or write-only
+    OpenMode ->
+    -- |Just x => creates the file with the given modes, Nothing => the file must exist.
+    Maybe FileMode ->
+    -- | Append, exclusive, etc.
+    OpenFileFlags ->
+    IO Fd
+open_ str how maybe_mode OpenFileFlags{..} =
+    Fd <$> c_open str all_flags mode_w
+  where
+    all_flags = creat .|. flags .|. open_mode
+
+    (creat, mode_w) = case maybe_mode of
+        Nothing -> (0, 0)
+        Just x -> (o_CREAT, x)
+
+    flags =
+        (if append then o_APPEND else 0)
+            .|. (if exclusive then o_EXCL else 0)
+            .|. (if noctty then o_NOCTTY else 0)
+            .|. (if nonBlock then o_NONBLOCK else 0)
+            .|. (if trunc then o_TRUNC else 0)
+
+    open_mode = case how of
+        ReadOnly -> o_RDONLY
+        WriteOnly -> o_WRONLY
+        ReadWrite -> o_RDWR
+
+foreign import capi unsafe "HsUnix.h open"
+    c_open :: CString -> CInt -> CMode -> IO CInt
+
+openRaw ::
```

Comment:
most of the `*Raw` functions were taken from the [`System.Posix.IO`](https://hackage.haskell.org/package/unix-2.8.0.0/docs/System-Posix-IO.html) module, but needed to be adapted to return the error as K's `IOError` rather than throwing the error within Haskell.

Reply:
Same as above